src/TRestAxionGenericOptics.cxx

```diff
@@ -133,6 +133,23 @@ void TRestAxionGenericOptics::InitFromConfigFile() {
     this->Initialize();
 }
 
+///////////////////////////////////////////////
+/// \brief This calculates the interaction Point between the X-Ray photons path and a mirror in the predetermined layer
+///
+TVector3 TRestAxionGenericOptics::GetInteractionPoint(const TVector3& pos, const TVector3& dir) {
+    Int_t layer = TRestAxionOptics::GetEntranceShell(pos, dir);
+    Double_t r1 = fShellsRadii[layer].second;
+    Double_t angle = fShellsAngle[layer];
+    Double_t xSep = fShellsSep[layer];
+    Double_t fLength = TRestAxionOptics::GetMirrLength();
+    // Double_t distMirrors = 0 for first stack and (fLength + xSep) * cos(angle) for the second // distance of the mirror should be added in for second stack 
+    Double_t r2 = r1 - fLength * sin(angle);
+    // r1 = sqrt((pos[0] + s * dir[0]) * (pos[0] + s * dir[0]) +
+    //    (pos[1] + s * dir[1]) * (pos[1] + s * dir[1])) -
+    //  ((r2 - r1) * (pos[2] + s * dir[2] - distMirrors) / (cos(angle) * fLength))  //needs to be solved for s, maybe by incresing s for some value and comparing this to r1
```

Comment:
Maybe someone knows a better way in c++ to get s from this equation than to increase s slowly and comparing r1 to the rest of the term.

Reply:
All values except for s are known

Reply:
With ROOT you may define a `TF1` object, whichs is a maths function, then use the `Solver` to get the roots of that function. I found this tutorial with a quick search:

https://www.youtube.com/watch?v=7VgmFpxzK9M

Reply:
You will likely need to add the `#include <TF1.h>` header on your cxx

Reply:
I didn't look to the details, but I would expect the input position is where the particle hits the entrace plane, i.e. the position returned by `TRestAxionOptics::GetPositionAtEntrance`, while the output position is the position at the exit optics plane, i.e. that needs to be calculated by each specific optics process, the position should belong to the exit optics plane. And it must be implemented in a function named `TRestAxionGenericOptics::GetPositionAtExit`.

The input direction vector will be normalized to unity, and therefore it should be the output direction returned by `TRestAxionGenericOptics::GetDirectionAtExit`